### PR TITLE
logic_test: SET DATABASE is sticky

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

File: sql/logic_test.go [hunk 1]
```diff
@@ -172,6 +172,20 @@ func (t *logicTest) close() {
 
 // setUser sets the DB client to the specified user, creating it if needed.
 func (t *logicTest) setUser(user string) {
+	if t.db != nil {
```
Comment (MBerhault) [inline sql/logic_test.go]: it might be good to adjust the function comment.

**Comment by MBerhault** [comment]
LGTM

File: sql/logic_test.go [hunk 2]
```diff
@@ -172,6 +172,20 @@ func (t *logicTest) close() {
...
+		var dbName string
+
+		if err := t.db.QueryRow("SHOW DATABASE").Scan(&dbName); err != nil {
+			t.Fatal(err)
+		}
+
+		defer func() {
+			if _, err := t.db.Exec(fmt.Sprintf("SET DATABASE = %s", dbName)); err != nil {
```
Comment (petermattis) [inline sql/logic_test.go]: Use a parameter instead of `Sprintf`: `Exec("SET DATABASE = ", dbName)`.

Comment (vivekmenezes) [inline sql/logic_test.go]: I believe you should be able to add it to the Open() call below using database=dbName and it should work

File: sql/logic_test.go [hunk 3]
```diff
@@ -172,6 +172,20 @@ func (t *logicTest) close() {
...
 // setUser sets the DB client to the specified user, creating it if needed.
 func (t *logicTest) setUser(user string) {
+	if t.db != nil {
+		var dbName string
+
+		if err := t.db.QueryRow("SHOW DATABASE").Scan(&dbName); err != nil {
+			t.Fatal(err)
+		}
+
+		defer func() {
```
Comment (petermattis) [inline sql/logic_test.go]: Should add a comment about what this is doing: `When switching to a new user, propagate the existing DATABASE setting to the new connection`.

**Comment by petermattis** [comment]
LGTM

File: sql/logic_test.go (hunk 3 above)
Comment (tamird) [inline sql/logic_test.go]: DONE

File: sql/logic_test.go (hunk 2 above)
Comment (tamird) [inline sql/logic_test.go]: @petermattis parameter substitution doesn't work in `SET DATABASE`:
```diff
diff --git a/sql/logic_test.go b/sql/logic_test.go
index 87e1e6b..fe355f7 100644
--- a/sql/logic_test.go
+++ b/sql/logic_test.go
@@ -181,7 +181,7 @@ func (t *logicTest) setUser(user string) {
 
 		defer func() {
 			// Propagate the DATABASE setting to the newly-live connection.
-			if _, err := t.db.Exec(fmt.Sprintf("SET DATABASE = %s", dbName)); err != nil {
+			if _, err := t.db.Exec("SET DATABASE = $1", dbName); err != nil {
 				t.Fatal(err)
 			}
 		}()
```
results in:
```
--- FAIL: TestLogic (0.13s)
	logic_test.go:185: syntax error at or near "1"
		SET DATABASE = $1
		               ^

FAIL
exit status 1
```

@vivekmenezes: won't work when the client is being returned from the map.

**Comment by vivekmenezes** [comment]
LGTM

**Pull request merged**